docs/rfcs/20220829-distributed-dynamic-routing.md

```diff
@@ -0,0 +1,224 @@
+
+
+#### Shard Migrate
+
+- Leader Shard：
+    1. First, CeresMeta adjusts the routing relationship and create a new follower shard on the node.
+    1. Follower shard notifies CeresMeta after initialization.
+    1. CeresMeta started the process of leader switch, switched the leader to the newly created follower shard.
+    1. Close the original leader shard after the leader switch is completed.
+- Follower Shard：
+    1. Adjust shard node mapping to directly assign this shard to another node.
+
+#### Shard Split
+
+1. CeresMeta creates a new shard and distributes it to the node with the lowest load.
+1. After the new shard is created, the table routing relationship in the shard is processed, and a part of the table is split to the new shard according to certain rules.
+
+Pending Problem：
+
+1. How to decide which tables need to be split into new shards?
+
+### Processing flow
+#### Cluster registration and initialization
+
+1. Meta received the request for cluster registration, create cluster meta info and save in etcd, set the ClusterTopologyState is EMPTY.
+1. Waiting for node regisration, When the number of registered nodes has reached minimum number required by the cluster, the cluster initialization process will start.
+1. Initialize the shard routing topology and distribute the shard evenly on each node.
+1. Set the ClusterTopologyState to STABLE, initialization complete, after that, the cluster can normally provide external services.
+
+#### Create Table
+
+1. The client initiates a create table request. In RPC mode, it will access the specified endpoint returned by CeresmMeta. In HTTP mode, it will randomly access an available endpoint.
+1. After receiving the request, the node will complete the table creation operation and notify CeresMeta. All subsequent reads and writes to this table will be processed by the node. CeresMeta will modify the routing relationship between the table and the shard and associate this table with the shard of the current node.
+1. The routing relationship of the new table has been registered. Now the table can be read and written to the outside normally.
+
+#### Write / Read Table
+
+1. Client accesses the endpoint specified by ceresmeta through RPC. Because the client caches the routing relationship, it may be expired.
+1. After receiving the request, the node needs to decide whether to process the request normally according to the latest routing table.
+    - Shard is in normal status and handles requests normally.
+    - The current shard is no longer on this node. The node returns an expiration flag. The client re initiates the route request to refresh the routing relationship.
+
+# Drawbacks
+#### Table Split
+CeresDB's current minimum scheduling unit is table. It does not support splitting or merging table. It can only migrate tables between nodes. In extreme cases, a single table may be too large for a single node to handle.
+
+# Alternatives
+#### Remove Shard
```

Comment:
Shard is a key design concept for the case where the number of the tables is large to avoid the performance bottleneck from CeresMeta.

Without shard, CeresMeta has to process tables directly, and it sounds convenient and flexible. However, in some cases, it will lead to a waste, e.g. every time the topology of cluster changes, all the CeresDB instances have to pull all the tables belonging to themselves from CeresMeta even if there is no any relevant table's migration.